### PR TITLE
[Merged by Bors] - refactor: downgrade `linear_combination`

### PR DESCRIPTION
**Pull request by hrmacbeth**
This PR drops support from several operations from the `linear_combination` tactic:
- multiplication of two hypotheses
- division by a hypothesis
- inversion of a hypothesis
- addition or subtraction of a constant and a hypothesis

The first three operations are removed in order to make `linear_combination` perform only genuinely linear operations; this makes the tactic more predictable and also allows for more flexibility in refactors/reimplementations. The last operation is removed because, after of the removal of the other three, it is redundant (adding a constant to both sides of an equation doesn't change the difference of the two sides).

The old operations are still available using the `congr(...)` term elaborator or using the legacy `linear_combination'` tactic.

A side effect that be seen in `Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean` is that elaboration changes slightly; previously `⁻¹` was one of the supported operations *on hypotheses*, so `⁻¹` got handled differently *in constants* too.  In that example, previously the atom which got pre-elaborated was `C'.u`; now the atom which gets pre-elaborated is `C'.u⁻¹`.

Discussion on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Narrowing.20the.20scope.20of.20.60linear_combination.60)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/LinearCombination.lean
Comment (eric-wieser) [inline Mathlib/Tactic/LinearCombination.lean:199]: Losing support for semirings feels like a bit of a shame; would it make sense to combine the syntaxes, but fallback to this approach internally if no ring can be found?

Comment (hrmacbeth) [reply Mathlib/Tactic/LinearCombination.lean:199]: I really think the tactic is *improved* by removing this from the scope.  I laid out my thoughts here
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Narrowing.20the.20scope.20of.20.60linear_combination.60/near/462845544
(and let's continue the discussion there rather than here), but for the record, let me write precisely what I think that scope should be:

`linear_combination` is a tactic which reduces problems of the form
```
h₁ : a₁ = b₁
h₂ : a₂ = b₂
h₃ : a₃ = b₃
⊢ A = B
```
to the task of showing that 
```
(f a₁ a₂ a₃ - f b₁ b₂ b₃) - (A - B)
```
equals zero (and then tries a standard normalization procedure to check that).  I believe the scope of the tactic should be limited to operations `f` for which this is the same thing as
```
f (a₁ - b₁) (a₂ - b₂) (a₃ - b₃) - (A - B)
```
Namely, **linear** operations `f`.

**Comment by urkud** [comment]
Please describe the change of scope in the commit message (e.g., by copy+pasting relevant paragraphs from the Zulip thread).

**Comment by hrmacbeth** [comment]
This is something that won't be settled until #16103 is decided about.

File: Mathlib/Tactic/LinearCombination.lean
Comment (hrmacbeth) [reply Mathlib/Tactic/LinearCombination.lean]: To summarize where we landed after the Zulip discussion and some follow-up PRs:
- the PR #16103 added *native* support for cancellative semirings in `linear_combination`;
- the PR #16166 preserved (near-) backward compatibility (in particular the original syntax and the rudimentary support for non-cancellative semirings) by duplicating `linear_combination` to a tactic `linear_combination'` which will not be further modified.

Comment (urkud) [reply Mathlib/Tactic/LinearCombination.lean]: I don't think that removing support for `Nat` makes the tactic better. I would prefer the tactic to swap LHS and RHS as needed to avoid `-`.

Comment (hrmacbeth) [reply Mathlib/Tactic/LinearCombination.lean]: @urkud I'm not sure what you mean here -- as of #16103 Nat is supported natively (and `-` *means* swapping).  Can you say a bit more about your concern?

Comment (urkud) [reply Mathlib/Tactic/LinearCombination.lean]: I wrote my comment before the description of the scope change was added. Please ignore it.

File: Mathlib/Tactic/LinearCombination.lean
Comment (eric-wieser) [inline Mathlib/Tactic/LinearCombination.lean:65]: nit: I'd argue that if it were agnostic to them, this would be a warning not an error. Throwing an error doesn't sound agnostic to me!

Comment (hrmacbeth) [reply Mathlib/Tactic/LinearCombination.lean:65]: OK, let me know if you have alternative wording that still accomplishes the goal of dropping this functionality!

**Comment by hrmacbeth** [comment]
I believe the Zulip discussion reached consensus in favour of this change, so I will remove that tag.

File: Mathlib/Tactic/LinearCombination.lean
Comment (eric-wieser) [reply Mathlib/Tactic/LinearCombination.lean:65]: Either
```suggestion
    | _ , _ => throwError "'linear_combination' does not support addition of constants as it would have no effect"
```
or
```suggestion
    | .proof p, .const _ | .const _, .proof p =>
      logWarning "'linear_combination' is agnostic to the addition of constants, so this has no effect"
      pure (.proof p)
```
where the second one allows users to ignore the warning

**Comment by eric-wieser** [comment]
Just to check: there is no docstring change here because the feature was never documented in the first place? Or was the docstring pre-emptively updated in a previous PR?

**Comment by hrmacbeth** [comment]
> Just to check: there is no docstring change here because the feature was never documented in the first place? Or was the docstring pre-emptively updated in a previous PR?

Right, the docstring dates from the Lean 3 tactic, whereas the features were added in the port.

File: Mathlib/Tactic/LinearCombination.lean
Comment (hrmacbeth) [reply Mathlib/Tactic/LinearCombination.lean]: I like the second one and have implemented it.  Let's revisit this if needed when the other planned refactors come.

File: Mathlib/Tactic/LinearCombination.lean
Comment (eric-wieser) [inline Mathlib/Tactic/LinearCombination.lean:90]: Can you add a test for this?

**Review by joneugster — CHANGES_REQUESTED** [review]
I think it would be reasonable if you updated the docstrings of `linear_combination` and `linear_combination'`:

> `both`: If the `normalize` field of the configuration is set to false

I think this should be changed to using `(norm := skip)`, right? (that's not directly relevant here, but I tripped over it when looking at the one situation above you used `linear_combination'`)

> `linear_combination'`: Note: There is also a similar tactic `linear_combination` (no prime); this version is provided for backward compatibility.

At least mention the difference implemented in this PR.

File: Mathlib/Tactic/LinearCombination.lean
Comment (joneugster) [reply Mathlib/Tactic/LinearCombination.lean:90]: I have to say, debugging the example above I found this message not immensely helpful. Is it possible that the error message mentions the term it is upset about?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean
Comment (hrmacbeth) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:504]: The most direct translation would be to use the `congr(...)` term elaborator:
```lean
  linear_combination' (norm := (rw [negY]; ring1)) Q z ^ 6 * (equation_iff P).mp hP
    - P z ^ 6 * (equation_iff Q).mp hQ + congr($hx ^ 3) + W'.a₂ * P z ^ 2 * Q z ^ 2 * congr($hx ^ 2)
    + (W'.a₄ * P z ^ 4 * Q z ^ 4 - W'.a₁ * P y * P z * Q z ^ 4) * hx
```
But in general, a "nonlinear" linear combination has a code smell to me.  For example, I would find it more elegant to solve this example in this way, using a purely "linear" linear combination and avoiding the nonstandard discharger:
```lean
  rw [equation_iff, negY] at *
  linear_combination Q z ^ 6 * hP - P z ^ 6 * hQ
    + (Q z ^ 4 * P z ^ 4 * W'.a₄ - P y * Q z ^ 4 * P z * W'.a₁ 
      + Q z ^ 2 *  P z ^ 2 * (P z ^ 2 * Q x + Q z ^ 2 * P x) * W'.a₂ 
      + (P z ^ 4 * Q x ^ 2 + Q z ^ 2 * P z ^ 2 * Q x * P x + Q z ^ 4 * P x ^ 2)) * hx
```

Comment (hrmacbeth) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:504]: In any case, did you write these variants to check your understanding of the "story" of this PR, or are they intended as genuine suggested changes?  The approach I have taken on these two examples which are hit by the downgrade -- switching them over to `linear_combination'`, to serve as a record of the scope of functionality which has been removed --  was requested by @digama0:
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Narrowing.20the.20scope.20of.20.60linear_combination.60/near/465256620

File: Mathlib/Tactic/LinearCombination.lean
Comment (hrmacbeth) [reply Mathlib/Tactic/LinearCombination.lean]: I added tests, and I localized the error to the offending `*` or `/` symbol -- what do you think?

**Comment by hrmacbeth** [comment]
> I think it would be reasonable if you updated the docstrings of `linear_combination` and `linear_combination'`:
>
> > `linear_combination'`: Note: There is also a similar tactic `linear_combination` (no prime); this version is provided for backward compatibility.
> 
> At least mention the difference implemented in this PR.

I have done this (98579c8d49852d5ae861b3f892635ecb1847e8a5).  Just flagging that I don't necessarily think that it is sustainable in the long term to continue to document the differences between the two: in the future I hope to extend the scope of `linear_combination` quite substantially.

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (eric-wieser) [inline Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:253]: Why did this end up changing?

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean
Comment (joneugster) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:504]: Thanks for the explanation, your snippet is indeed much more elegant!

So my suggestion is to just replace these two uses of `linear_combination'` with the solution you just wrote out, since thats nicer than the original anyways.

But also, I think the PR can be merged regardless if you do that or not :)

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (joneugster) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:253]: I think I looked at this earlier and from top of my head:

Previously the tactic would parse the `\^-1` before the term `C.u` was parsed (and translate it into using one of the theorems that got removed) but now it's actually the normal term parser parsing the entire `C.u\^-1`.

Therefore, now it thinks the inverse should be taken in `R` instead of in `R\^x` without the coersion hint (which is btw also displayed in the goal). This seems like normal Lean behaviour so the coersion hints are in my opinion expected there.

**Comment by joneugster** [comment]
Thank you heather for addressing the comments! I'm conviced now that this is a positive change overall, and I'm looking forward to see more changes to the new `linear_combination`

Maybe Eric can "maintainer merge" once happy :)

File: Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean
Comment (hrmacbeth) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean:504]: If you don't mind either way, then let's go with the status quo, since Mario preferred it!

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (hrmacbeth) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:253]: Previously `⁻¹` was one of the supported operations *on hypotheses*, so `⁻¹` got handled differently *in constants* too: the atom which got pre-elaborated was `C'.u`.  Now the atom which gets pre-elaborated is `C'.u⁻¹`.  (Though I don't actually see why it makes a difference here!)

**Review by joneugster — APPROVED** [review]

File: Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean
Comment (eric-wieser) [reply Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean:253]: Thanks, I've copied @hrmacbeth's version into the PR description.

**Comment by joneugster** [comment]
I think since @eric-wieser marked the last conversation as resolved, I can suggest

maintainer merge

Thanks for the PR!

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**
